Components: Use Button component in app-promo

sarayourfriend opened the pull request:
#### Changes proposed in this Pull Request

* Replace usage of `button` and anchor tags with `Button` component in `app-promo`

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Add an early return to `client/reader/sidebar/promo.jsx`'s `shouldRenderAppPromo` to return `true` (it's too difficult to coordinate a legitimate usage). Then open the reader sidebar and you'll see the app promo at the bottom. Make sure to force a mobile and desktop one to happen.

## Before:

<img width="1456" alt="Captura de Tela 2020-09-01 às 12 50 38" src="https://user-images.githubusercontent.com/24264157/91900199-2d9e2100-ec53-11ea-8f0f-b15cd1199a94.png">

## After:

<img width="1456" alt="Captura de Tela 2020-09-01 às 12 54 30" src="https://user-images.githubusercontent.com/24264157/91900205-30991180-ec53-11ea-8413-9a79d2a8f99c.png">


Related to #45259

sarayourfriend committed:
Components: Use Button component in app-promo

tyxla reviewed (changes requested):
Thanks for this one, I think it's a much better approach for `button`->`Button` that is used basically all throughout Calypso to go granular vs all in one PR.

There are a couple of small visual changes this PR introduces that we should take care of before shipping it:

### Line height is different (larger)

Before:
![](https://cldup.com/3Z6e3zVv5U.png)

After:
![](https://cldup.com/iedCCP1ji7.png)

### Hover state color is different

Before:
![](https://cldup.com/6E6S535sN4.png)

After:
![](https://cldup.com/Sj1kmJ8png.png)

Also, there is a failing test.

> Add an early return to `client/reader/sidebar/promo.jsx`'s `shouldRenderAppPromo` to return `true` (it's too difficult to coordinate a legitimate usage). Then open the reader sidebar and you'll see the app promo at the bottom. Make sure to force a mobile and desktop one to happen.

FWIW, this can be done with a simple prop change at the Components section in the Chrome dev tools. That makes it significantly easier to test 😉

tyxla commented on client/blocks/app-promo/style.scss:
```diff
@@ -19,8 +19,8 @@
 .app-promo .app-promo__dismiss {
 	position: absolute;
 	right: 17px;
-	top: 12px;
```
Right now, there's a difference between production and this PR as to where the dismiss button appears (vertically) - it's a bit subtle:

![](https://cldup.com/iCw6VBpJss.png) ![](https://cldup.com/lC32Gpy-NC.png)

Bringing this line back but changing it to `top: 3px` does the job for me.

tyxla reviewed (changes requested):
Sorry for being picky, but there's a subtle visual difference we could fix easily.

Also, there is a failing test we should fix before 🚢 .

Other than that, I think this PR should be good to go - it works great in my tests.

sarayourfriend committed:
Components: Fix app-promo test

Now that we're rendering a react component instead
of the regular HTML tag, shallow isn't going to
render the children, so we need to use mount instead.

tyxla reviewed (approved):
LGTM 🚢  Thanks!

sarayourfriend committed:
Use WordPress button

saramarcondes commented:
@tyxla I've revisited this PR using WP components rather than `@automattic/components`. Can you let me know what you think. I think this is desirable based on the conversation in p9oQ9f-k2-p2.

tyxla reviewed (changes requested):
Thanks @saramarcondes, using `@wordpress/components` here is a great idea! It actually works great and I think we're pretty close!

I have a couple of concerns with this:

1. The `async-load-reader-sidebar` chunk size seems to have grown twice. This is odd - is it because we're using a new component? Shouldn't that component live somewhere else?
2. There is a visual glitch for the instances that end up with a `<button>`:

Production:
![](https://cldup.com/GMDa5qDPLs.png)

This branch:
![](https://cldup.com/Wkl2A3cJU0.png)

Other than that, I think we should start working in that direction and this is a great start! 🙌

saramarcondes commented:
@tyxla I wonder if tree-shaking isn't working with `@wordpress/components`? In any case, we could async-load the `AppPromo` component I think to offset that, right?

saramarcondes commented:
> There is a visual glitch for the instances that end up with a <button>:

@tyxla I can't reproduce that on Firefox or Chrome. Did you do anything specific to get that result?

sarayourfriend committed:
Async load AppPromo in reader sidebar

sarayourfriend committed:
Async load reader sidebar promo

tyxla reviewed (approved):
I don't know what you did, but I can no longer reproduce the visual glitch from before 🤷‍♂️. Maybe our ongoing style migration fixed that silently? 😛  

I've left a couple of comments that I'd prefer to have addressed before we ship. Other than those, this LGTM, and I think we should ship it 👍  This is exciting, we're making history here!

We can optimize the size impact further once we have multiple usages of `@wordpress/components` in Calypso.

saramarcondes commented on client/blocks/app-promo/index.jsx:
```diff
@@ -31,26 +36,26 @@ import './style.scss';
-			message: 'Fast, distraction-free WordPress.com — download the desktop app.',
-			type: 'desktop',
-		},
+		// {
```
Whoops! 😅

griffbrad commented on client/reader/sidebar/promo.jsx:
```diff
@@ -23,7 +23,7 @@ export const ReaderSidebarPromo = ( { currentUserLocale, shouldRenderAppPromo }
 			{ shouldRenderAppPromo && (
 				<div className="sidebar__app-promo">
-					<AppPromo location="reader" locale={ currentUserLocale } />
+					<AsyncLoad require="blocks/app-promo" location="reader" locale={ currentUserLocale } />
```
Should we make that the default? It seems like it comes up a lot when we’re loading components async.

tyxla replied:
There's a good reason to have this placeholder when something's dynamically loading - often it is useful to know that something is loading. But for components that may not display anything, it might be a false alarm - one that could be confusing to the user.
